R/generate_eval_data.R

```diff
@@ -0,0 +1,110 @@
+#' Generate evaluation data for a hub
+#'
+#' @param hub_path A path to the hub.
+#' @param config_path A path to a yaml file that specifies the configuration
+#' options for the evaluation.
+#' @param out_path The directory to write the evaluation data to.
+#' @param oracle_output A data frame of oracle output to use for the evaluation.
+#'
+#' @export
+generate_eval_data <- function(hub_path,
+                               config_path,
+                               out_path,
+                               oracle_output) {
+  config <- read_config(hub_path, config_path)
+  for (target in config$targets) {
+    generate_target_eval_data(hub_path, config, out_path, oracle_output, target)
+  }
+}
+
+
+#' Generate evaluation data for a target
+#'
+#' @inheritParams generate_eval_data
+#' @param config The configuration options for the evaluation.
+#' @param target The target to generate evaluation data for. This is one object
+#' from the list of targets in the config, with properties "target_id",
+#' "metrics", and "disaggregate_by".
+#'
+#' @noRd
+generate_target_eval_data <- function(hub_path,
+                                      config,
+                                      out_path,
+                                      oracle_output,
+                                      target) {
+  target_id <- target$target_id
+  metrics <- target$metrics
+  disaggregate_by <- target$disaggregate_by
+  eval_windows <- config$eval_windows
+
+  task_groups_w_target <- get_task_groups_w_target(hub_path, target_id)
+  metric_name_to_output_type <- get_metric_name_to_output_type(task_groups_w_target, metrics)
+
+  for (eval_window in eval_windows) {
+    model_out_tbl <- load_model_out_in_window(hub_path, target$target_id, eval_window)
+
+    # overall scores, not disaggregated by a task id variable
+    get_and_save_scores(
+      model_out_tbl = model_out_tbl,
+      oracle_output = oracle_output,
+      metric_name_to_output_type = metric_name_to_output_type,
+      target_id = target_id,
+      window_name = eval_window$window_name,
+      by = NULL,
+      out_path = out_path
+    )
+
+    # scores disaggregated by a task id variable
+    for (by in disaggregate_by) {
```

Comment:
You can generate the overall scores along with the disaggregate scores by appending `NULL` to the list of disaggregations in this for loop.

```suggestion
    # calculate overall scores followed by scores disaggregated by a task ID variable.
    for (by in c(list(NULL), disaggregate_by)) {
```

Reply:
Neat trick with `list(NULL)`

Reply:
Yeah, it's kind of an advanced R technique that requires you to know that if you combine a list with a vector, the vector is coerced to a list as shown in the examples of [the `c` function documentation](https://stat.ethz.ch/R-manual/R-patched/library/base/html/c.html):

```
## append to a list:
ll <- list(A = 1, c = "C")
## do *not* use
c(ll, d = 1:3) # which is == c(ll, as.list(c(d = 1:3)))
## but rather
c(ll, d = list(1:3))  # c() combining two lists
```

That being said.... I'm now realizing that it should be more explicit that `disaggregate_by` is coerced to a list.